Sources/PackagePlugin/Implementation.swift

```diff
@@ -0,0 +1,132 @@
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+@_implementationOnly import Foundation
+
+// The way in which SwiftPM communicates with the package plugin is an im-
+// plementation detail, but the way it currently works is that the plugin
+// is compiled (in a very similar way to the package manifest) and then run in
+// a sandbox. Currently it is passed the JSON encoded input structure as the
+// last command line argument; however, it this will likely change to instead
+// read it from stdin, since that avoids any command line length limitation.
+// Any generated commands and diagnostics are emitted on stdout after a zero
+// byte; this allows regular output, such as print statements for debugging,
+// to be emitted to SwiftPM verbatim. SwiftPM tries to interpret any stdout
+// contents after the last zero byte as a JSON encoded output struct in UTF-8
+// encoding; any failure to decode it is considered a protocol failure. The
+// exit code of the compiled plugin determines success or failure (though
+// failure to decode the output is also considered a failure to run the ex-
+// tension).
+
+extension Plugin {
+
+    public static func main() throws {
+        // Look for the input JSON as the last argument of the invocation.
+        guard let inputData = ProcessInfo.processInfo.arguments.last?.data(using: .utf8) else {
+            fputs("Expected last argument to contain JSON input data in UTF-8 encoding, but didn't find it.", stderr)
+            output.diagnostics.append(Diagnostic(severity: .error, message: "Expected last argument to contain JSON input data in UTF-8 encoding, but didn't find it.", file: nil, line: nil))
+            exit(1)
+        }
+ 
+        // Decode the input JSON into a plugin context.
+        let context: TargetBuildContext
+        do {
+            let decoder = JSONDecoder()
+            context = try decoder.decode(TargetBuildContext.self, from: inputData)
+        }
+        catch {
+            fputs("Couldn't decode input JSON (reason: \(error)", stderr)
+            output.diagnostics.append(Diagnostic(severity: .error, message: "\(error)", file: nil, line: nil))
+            exit(1)
+        }
+
+        // Instantiate the plugin. For now there are no parameters, but this is where we would set them up, most likely as properties of the plugin.
+        let plugin = self.init()
+        
+        // Invoke the appropriate protocol method based on the action.
+        switch context.pluginAction {
+        case .createBuildToolCommands:
+            // Check that the plugin conforms to `BuildToolPlugin`, and get the commands.
+            guard let plugin = plugin as? BuildToolPlugin else { throw PluginDeserializationError.malformedInputJSON("...") }
```

Comment:
better error?

Reply:
Ah, right, missed this one.  :)  Thanks!